002-es6-modules.md

```diff
@@ -0,0 +1,274 @@
+    - A hook for when an import is exactly performed.
+    
+* [CreateImportBinding](https://tc39.github.io/ecma262/#sec-createimportbinding)
+    - A means to create a shared binding (variable) from an export to an import. Required for the "live" binding of imports.
+
+* [ModuleNamespaceCreate](https://tc39.github.io/ecma262/#sec-modulenamespacecreate)
+    - Provides a means of creating a list of exports manually, used so that CommonJS `module.exports` can create ModuleRecords that are prepopulated.
+
+---
+
+### [WhatWG Loader](https://github.com/whatwg/loader)
+
+---
+
+Discusses the design of module metadata in a [Registry](https://whatwg.github.io/loader/#registry). All actions regarding the Registry are synchronous.
+
+**NOTE** It is not Node's intent to implement the asynchronous pipeline in the Loader specification. There is discussion about including a synchronous pipeline in the specification as an addendum.
+
+
+### [Summary Video on Youtube](youtube.com/watch?v=NdOKO-6Ty7k)
+
+## Additional Structures Required
+
+### DynamicModuleRecord
+
+A Module Record that presents a view of an Object for its `[[Namespace]]` rather than coming from an environment record.
+
+The list of exports is frozen upon construction. No new properties may be added. No properties may be removed.
+
+## Algorithm
+
+When `require()`ing a file.
+
+1. Determine if file is ES6 or CommonJS (CJS).
+2. If CJS
+	1. Evaluate immediately
+	2. Produce a DynamicModuleRecord from `module.exports`
+3. If ES6
+	1. Parse for `import`/`export`s and keep record, in order to create bindings
+	2. Gather all submodules by performing `require` recursively
+		* See circular dep semantics below 
+	3. Connect `import` bindings for all relevant submodules (see [ModuleDeclarationInstantiation](https://tc39.github.io/ecma262/#sec-moduledeclarationinstantiation))
+	4. Evaluate
+
+## Semantics
+
+### Determining if source is an ES6 Module
+
+If a module can be parsed outside of the ES6 Module goal, it will be treated as a [Script](https://tc39.github.io/ecma262/#sec-parse-script). Otherwise it will be parsed as [ES6](https://tc39.github.io/ecma262/#sec-parsemodule).
+
+In pseudocode, it looks somewhat like:
+
+```
+try {
+  v8::Script::Compile
+}
+catch (e) {
+  v8::Script::CompileModule
+}
+```
+
+V8 may or may not choose to use parser fallback to combine this into one step.
+
+### CommonJS consuming ES6
+
+#### default exports
+
+ES6 modules only ever declare named exports. A default export just exports a property named `default`.
+
+Given
+
+```javascript
+let foo = 'my-default';
+export default foo;
+```
+
+```javascript
+require('./es6');
+// {default:'my-default'}
+```
+
+#### read only
+
+The objects create by an ES6 module are [ModuleNamespace Objects](https://tc39.github.io/ecma262/#sec-module-namespace-exotic-objects).
+
+These have `[[Set]]` be a no-op and are read only views of the exports of an ES6 module.
+
+### ES6 consuming CommonJS
```

Comment:
This thread surprises me greatly. The only reason ES6 modules are something worth using for many of us is because for the (likely very long) transition period to "everything is ES6 modules", the promise everyone's been sold on is that "it will be easy". I should be able to `import` CJS modules, and ideally (but not necessarily) `require` ES6 modules. I should be able to change my CJS module to an ES6 module without changing anything in consumers, and, ideally (but not necessarily) I should be able to change an ES6 module to CJS without changing anything in consumers. (Certainly I'm less concerned about moving backwards than I am about moving forwards).

Devs aren't interested in mixing `require()` and `import`. They want to use one or the other, all the time. Many already are, via Babel - they're *already* `import`ing CJS modules.

Just because the semantics *could* be different doesn't mean they must be. To use your example, most ES6 modules won't ever use the "live binding" semantic - ie, they'll never change their exports after they're set. Any module graph benefits will likely not be useful until entire subtrees are using ES6 modules.

In general, if ES6 modules pull a Python 3, the community will simply not adopt them. People will build wrappers that `import` modules and `module.exports =` them, which seems like it would destroy the usefulness of an ES6 module dep graph. Absent a clear adapter convention, people will publish modules that have two entry points instead of one - one for CJS, and one for ES6. All of this will create a lot of developer pain and annoyance, and given that ES6 imports are second-class, not first-class, it's likely that many will just stay with (or switch back to) `require`.

The majority of the ecosystem values backwards compatibility: with old node versions, old browsers, old JS engines, etc. I think perhaps that you may be vastly undervaluing the absolutely critical-ness of interop.